File: AndroidAppBaseLibrary/src/main/java/stark/android/appbase/recycleview/BaseRecycleViewAdapter.java
package stark.android.appbase.recycleview;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

public abstract class BaseRecycleViewAdapter<T> extends RecyclerView.Adapter<BaseRecycleViewHolder<T>> {

    protected List<T> items;

    protected LayoutInflater mInflater;

    public BaseRecycleViewAdapter(Context context) {
        mInflater = LayoutInflater.from(context);
    }

    public void setItems(List<T> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public void addItems(List<T> items) {
        items.addAll(items);
        notifyDataSetChanged();
    }

    @Override
    public BaseRecycleViewHolder<T> onCreateViewHolder(ViewGroup parent, int viewType) {
        final BaseRecycleViewHolder viewHolder = onCreateViewHolderDelegate(parent, viewType);
        return viewHolder;
    }

    public abstract BaseRecycleViewHolder<T> onCreateViewHolderDelegate(ViewGroup parent, int viewType);

    @Override
    public void onBindViewHolder(BaseRecycleViewHolder<T> holder, int position) {
        holder.bindTo(items.get(position));
    }

    @Override
    public int getItemCount() {
        return items == null ? 0 : items.size();
    }
}

File: AndroidAppBaseLibrary/src/main/java/stark/android/appbase/recycleview/BaseRecycleViewAdapter.kt
package stark.android.appbase.recycleview

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

abstract class BaseRecycleViewAdapter<T>(context: Context?) : RecyclerView.Adapter<BaseRecycleViewHolder<T>>() {

    internal var items: MutableList<T>? = null

    open var mInflater: LayoutInflater = LayoutInflater.from(context)

    fun setItems(items: MutableList<T>) {
        this.items = items
        notifyDataSetChanged()
    }

    fun getItems() :MutableList<T>?{
        return items
    }

    fun addItems(items: MutableList<T>) {
        items.addAll(items)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecycleViewHolder<T> {
        return onCreateViewHolderDelegate(parent, viewType)
    }

    abstract fun onCreateViewHolderDelegate(parent: ViewGroup, viewType: Int): BaseRecycleViewHolder<T>

    override fun onBindViewHolder(holder: BaseRecycleViewHolder<T>, position: Int) {
        holder.bindTo(items!![position])
    }

    override fun getItemCount(): Int {
        return if (items == null) 0 else items!!.size
    }
}
